Bluetooth: has: Ensure that notifications are sent

MariuszSkamra opened the pull request:
This ensures active preset index, HAS features and partially control point notifications are sent.
The code retries sending notifications in case of lack of available buffers.
Partially addresses #57454

MariuszSkamra committed:
Bluetooth: audio: has: Fix checking wrong flag

This fixes testing and clearing features flag.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Fix indentation

This fixes code indentation.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Defer notifications to sysworkq

Defer sending the features, active index, preset list and preset read
response to sysworkq and retry sending in case failed due to buffers not
available at the moment.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Don't revert features value on error

This removes revering the features value back to previous state if work
submission failes. Even if it fails it indicates an internal sysworkq
issue, so even retry won't help. The client can read the features value
anyway, thus it's sane to just log an error in such case.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: has: Add test Preset Changed Offline Behavior

Verify that a HAS Server IUT sends changed characteristic notifications
or indications when the Lower Tester reconnects.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Fix sending notifications on reconnect

This fixes missing setting of FLAG_CONTROL_POINT_NOTIFY flag that
indicate whether submit control_point_work. In case the there are more
indications/notifications to sent (is_last flag is unset), the
FLAG_CONTROL_POINT_NOTIFY shall be set to resubmit control_point_work.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Factor out set_preset_availability function

This moves common code to set_preset_availability function to be called
from bt_has_preset_available and bt_has_preset_unavailable.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Refactor preset list to single-linked list

This refactors the preset list to use sys_slist API. There have been
various issues seen while iterating presets, thus it's more save to use
well-defined and tested sys_slist API.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on tests/bsim/bluetooth/audio/src/has_client_test.c:
```diff
@@ -227,25 +244,426 @@ static void test_main(void)
+static struct bt_gatt_subscribe_params features_sub;
+static struct bt_gatt_subscribe_params active_preset_index_sub;
+static struct bt_gatt_subscribe_params control_point_sub;
+static int notify_received_mask;
```
Any reason why the mask is an `int`?

Thalley commented on tests/bsim/bluetooth/audio/test_scripts/has.sh:
```diff
@@ -24,3 +25,18 @@ Execute ./bs_2G4_phy_v1 -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} \
 
 wait_for_background_jobs
+
+printf "\n\n Running Preset Changed Offline Behavior test \n\n"
+
+SIMULATION_ID="has_offline_behavior"
+
+Execute ./bs_${BOARD}_tests_bsim_bluetooth_audio_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -testid=has_offline_behavior -rs=24
+
+Execute ./bs_${BOARD}_tests_bsim_bluetooth_audio_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=1 -testid=has_client_gatt -rs=46
+
+Execute ./bs_2G4_phy_v1 -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} \
+  -D=2 -sim_length=60e6 $@
+
+wait_for_background_jobs
```
Suggest to put this into a new script, (e.g. `has_offline.sh` so that it can run in parallel the "regular" HAS test

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -328,6 +301,89 @@ BT_CONN_CB_DEFINE(conn_cb) = {
+static void notify(struct has_client *client, enum has_notify notify)
+{
+	if (client != NULL) {
+		notify_work_reschedule(client, notify, K_NO_WAIT);
```
When calling `k_work_reschedule` with `K_NO_WAIT`, that effectively dismissed the earlier scheduled work, and attempts to schedule it without waiting. That means that if we had scheduled a workitem with `K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US)`, the delay will be dismissed and the notification will be attempted with K_NO_WAIT. 

I don't think it's a problem, but is this intentional? Or would it more sense to check if `k_work_delayable_is_pending` before scheduling with `K_NO_WAIT`?

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -601,74 +645,73 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
-		if (preset == NULL) {
-			return;
-		}
+	preset_foreach(client->preset_changed_index_next,
+			BT_HAS_PRESET_INDEX_LAST, preset_found, &preset);
```
Minor indentation fix

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -601,74 +645,73 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
+	}
 
-		is_last = next == NULL;
+	preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
+			preset_found, &next);
```
Minor indentation fix

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -433,57 +442,122 @@ static uint8_t preset_found(const struct has_preset *preset, void *user_data)
+{
+	struct has_preset *tmp, *prev = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&preset_list, tmp, node) {
+		if (tmp->index == preset->index) {
+			if (prev == NULL && wrap) {
+				sys_snode_t *node = sys_slist_peek_tail(&preset_list);
+
+				prev = CONTAINER_OF(node, struct has_preset, node);
+				if (prev == preset) {
+					return NULL;
+				}
+			}
+			break;
+		}
+		prev = tmp;
+	}
```
I have a hard time following this. 

If you want to find the prev, shouldn't it just be 

```c
	SYS_SLIST_FOR_EACH_CONTAINER(&preset_list, tmp, node) {
		if (SYS_SLIST_PEEK_NEXT_CONTAINER(tmp) == preset) {
			return tmp;
		}
	}
```

Why do we need to compare `index` or do wrapping?

Thalley replied:
Alternatively, we could consider using `dlist` if we want to iterate on the list in both directions

MariuszSkamra replied:
Thanks for `SYS_SLIST_PEEK_NEXT_CONTAINER`, will use it. For such simple use case it's not worth to use dlist IMO

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -433,57 +442,122 @@ static uint8_t preset_found(const struct has_preset *preset, void *user_data)
+	}
+
+	return next;
+}
```
Ditto: Why do we need to `wrap`?

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -397,19 +405,20 @@ static struct has_preset {
 
-/* Number of registered presets */
-static uint8_t has_preset_num;
+static sys_slist_t preset_list = SYS_SLIST_STATIC_INIT(&preset_list);
+static sys_slist_t preset_free_list = SYS_SLIST_STATIC_INIT(&preset_free_list);
```
Can you elaborate on why we need two lists?

MariuszSkamra replied:
`preset_free_list` is a pool of free preset entries, while `preset_free_list` is the actual list of presets

MariuszSkamra committed:
Bluetooth: has: Factor out bonded client persistent data

Some of the data shall be persistent across connections to bonded
clients. This includes notidication state flags that are used to
determine whether notify bonded client after reconnection.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -328,9 +305,97 @@ BT_CONN_CB_DEFINE(conn_cb) = {
+	struct has_client *client = CONTAINER_OF(dwork, struct has_client, notify_work);
+	int err;
+
+#if defined(CONFIG_BT_HAS_FEATURES_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->notify, NOTIFY_FEATURES) &&
+	    bt_gatt_is_subscribed(client->conn, FEATURES_ATTR, BT_GATT_CCC_NOTIFY)) {
+		err = bt_gatt_notify(client->conn, FEATURES_ATTR, &has.features,
+				     sizeof(has.features));
+		if (err == -ENOMEM) {
+			notify_work_reschedule(client, NOTIFY_FEATURES,
+					       K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+		} else if (err < 0) {
+			LOG_ERR("Notify features err %d", err);
+		}
+	}
+#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
```
Correct me if I am wrong, but it doens't seem like this block contains any data that isn't always available, in which case the `if defined` can be replaced with `if IS_ENABLED`

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -328,9 +305,97 @@ BT_CONN_CB_DEFINE(conn_cb) = {
+	}
+#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+
+#if defined(CONFIG_BT_HAS_PRESET_SUPPORT)
```
Looks like this may also be able to be guarded by `IS_ENABLED` instead

MariuszSkamra replied:
In this particular case, I cannot use IS_ENABLED, because the functions called below are undeclared

Thalley commented on tests/bsim/bluetooth/audio/src/has_client_test.c:
Why are we mostly re-implementing the HAS client again here? 

Couldn't we reuse a lot of the code from the actual implementation?

MariuszSkamra replied:
As HAS client unsubscribes from the notifications on disconnection, I cannot use it for offline behavior test cases

Thalley replied:
> As HAS client unsubscribes from the notifications on disconnection, I cannot use it for offline behavior test cases

Wouldn't it be easier to implement support for bonding in the client, than reimplementing a new HAS client that supports that? This just seems like much more effort than fixing the missing feature from the current HAS client :) 

But can be done in a separate PR.

Thalley commented on tests/bsim/bluetooth/audio/src/has_client_test.c:
```diff
@@ -229,25 +246,448 @@ static void test_main(void)
 	},
+#endif /* CONFIG_BT_HAS_CLIENT */
+	{
+		.test_id = "has_client_gatt",
```
The `_gatt` suffix isn't very descriptive. If you want to keep the name, please consider adding some text to `.test_descr`

Thalley commented on tests/bsim/bluetooth/audio/test_scripts/has_offline.sh:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2022 Codecoup
```
```suggestion
# Copyright (c) 2023 Codecoup
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -1117,9 +1117,12 @@ int bt_has_preset_unregister(uint8_t index)
-int bt_has_preset_available(uint8_t index)
+static int set_preset_availability(uint8_t index, bool available)
 {
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(struct bt_has_cp_hdr) +
+			      sizeof(struct bt_has_cp_preset_changed) + sizeof(uint8_t));
```
> Which btw, this isn't really legal in C99?

IIRC Zephyr moved to C11 a while ago

fredrikdanebjer replied:
It is actually the same in ISO 9899:2011 6.7.6.2:
> If the expression is a constant expression, it shall have a value greater than zero.

However, according to https://docs.zephyrproject.org/latest/develop/languages/c/index.html Zephyr does not target a specific C standard, it just claims that many features introduced in C99 and C11 is used. Since they are all over the place in Zephyr it has to be fine. Just saying, I prefer the other solution :D

Thalley replied:
You are right - The flexibly array (denoted by `[]` instead of `[0]`) is valid in C99 though, and the `[0]` seems to be a GCC extension (and also supported in LLVM). 

Nevertheless it's a common practice to use in Zephyr (searching for ` \w+\[0\];\n\}` yields 195 results, the majority of which are in the BT subsystem). It is probably something that should be replaced by `[]` or the newer bounded flexible arrays (https://people.kernel.org/kees/bounded-flexible-arrays-in-c)

Thalley replied:
Created https://github.com/zephyrproject-rtos/zephyr/issues/63845

Thalley replied:
Defining `sizeof(struct bt_has_cp_hdr) + sizeof(struct bt_has_cp_preset_changed) + sizeof(uint8_t)` as a single struct or value would still make sense to do in this PR IMO

Thalley replied:
Defining `sizeof(struct bt_has_cp_hdr) + sizeof(struct bt_has_cp_preset_changed) + sizeof(uint8_t)` as a single struct or value would still make sense to do in this PR IMO

MariuszSkamra replied:
Those structures are common for client and server. I would need to change the client code as well

Thalley replied:
Not suggesting to change any existing, but rather a new one that suits this. Not sure I understand the issue with doing so

MSEEHenrik commented on subsys/bluetooth/audio/has.c:
```diff
@@ -589,7 +687,7 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
 
 	generic_update = net_buf_simple_add(&buf, sizeof(*generic_update));
-	generic_update->prev_index = get_prev_preset_index(preset);
+	generic_update->prev_index = preset_get_prev_index(preset);
```
This will provide a wrapped index.  The first item in the list has to have a prevIndex of **0**.

MariuszSkamra committed:
Bluetooth: has: Fix preset list notifications after reconnection

This fixes sending proper Preset List notifications after
reconnection. The issue was observed when the last preset
known to the client has been removed.
As we do not hold the information about the deleted presets,
we need to use Generic Update procedure to:
 1. Notify the presets that have been removed in range
    (PrevIndex = current_preset_last, Index=previous_preset_last)
 2. Notify deletion of preset Index=previous_preset_last.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
Fixed preset list notifications after reconnection if the tail of the preset list have changed (the las one preset have been removed).

In such case:
```
/* The last preset known to the client has been removed.
	* As we do not hold the information about the deleted presets, we need to use
	* Generic Update procedure to:
	*   1. Notify the presets that have been removed in range
	*      (PrevIndex = current_preset_last, Index=previous_preset_last)
	*   2. Notify deletion of preset Index=previous_preset_last.
	*/
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -39,6 +37,8 @@ LOG_MODULE_REGISTER(bt_has, CONFIG_BT_HAS_LOG_LEVEL);
 	!BITS_CHANGED(_new_value, ((has.features & BT_HAS_FEAT_PRESET_SYNC_SUPP) != 0 ? 1 : 0))
 #define FEATURE_IND_PRESETS_UNCHANGED(_new_value) \
 	!BITS_CHANGED(_new_value, ((has.features & BT_HAS_FEAT_INDEPENDENT_PRESETS) != 0 ? 1 : 0))
+#define NOTIFY_AFTER_REBOOT \
+	(BIT(NOTIFY_ACTIVE_INDEX) | BIT(NOTIFY_PRESET_LIST) | BIT(NOTIFY_FEATURES))
```
I guess this is a mask, right? Suggest to add `_MASK` suffix, as the current name suggests either a verb or a boolean.

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -195,13 +199,72 @@ static struct has_client {
+{
+	struct client_context *context = NULL;
+
+	for (size_t i = 0; i < ARRAY_SIZE(contexts); i++) {
+		if (bt_addr_le_eq(&contexts[i].addr, BT_ADDR_LE_ANY)) {
+			context = &contexts[i];
+			break;
+		}
+	}
```
Could this be replaced by `context_find(BT_ADDR_LE_ANY)`?

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -387,70 +559,131 @@ static uint8_t preset_found(const struct has_preset *preset, void *user_data)
+		}
+	}
+
+	if (wrap) {
+		prev = SYS_SLIST_PEEK_TAIL_CONTAINER(&preset_list, prev, node);
+		if (prev == preset) {
+			return NULL;
+		}
+	}
```
I think I finally understood the `wrap` part: The list of presets are cyclical, right? So the prev of HEAD is TAIL, unless there is only 1 preset?

Why are we guarding this cyclical behavior here, when we don't in `preset_get_next`?

MariuszSkamra replied:
Because it won't be needed. The Generic Update PDU requires `prev_index` parameter only. However I'll move the logic to `preset_get_prev_index` function which is called in the context of Generic Update only

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -889,80 +1175,52 @@ static uint8_t handle_set_active_preset(struct net_buf_simple *buf, bool sync)
 
-	for (size_t i = 0; i < has_preset_num; i++) {
-		const struct has_preset *tmp = &has_preset_list[i];
+	if (active_preset == NULL) {
+		next = SYS_SLIST_PEEK_HEAD_CONTAINER(&preset_list, next, node);
+	} else {
+		next = preset_get_next(active_preset);
+	}
```
Minor, but you could consider implementing a `preset_get_head` for consistency

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -889,80 +1175,52 @@ static uint8_t handle_set_active_preset(struct net_buf_simple *buf, bool sync)
-			prev_available = tmp;
-			continue;
-		}
+	if (active_preset == NULL) {
+		prev = SYS_SLIST_PEEK_TAIL_CONTAINER(&preset_list, prev, node);
+	} else {
+		prev = preset_get_prev(active_preset, true);
+	}
```
Ditto with `preset_get_tail`

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -157,32 +156,37 @@ static struct bt_gatt_attr has_attrs[] = {
-#endif /* CONFIG_BT_HAS_PRESET_SUPPORT */
-#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+enum has_notify {
+	NOTIFY_ACTIVE_INDEX,
```
You have right, I'll revert it back

MariuszSkamra committed:
Bluetooth: audio: has: Fix missing memset of parameters

This fixes missing memset of parameters used for indications and/or
notifications.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -195,13 +197,67 @@ static struct has_client {
+static struct client_context *context_alloc(const bt_addr_le_t *addr)
+{
+	struct client_context *context;
+
+	context = context_find(BT_ADDR_LE_ANY);
```
```suggestion

	/* Free contexts has BT_ADDR_LE_ANY as the address */
	context = context_find(BT_ADDR_LE_ANY);
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -195,13 +197,67 @@ static struct has_client {
+	(void)k_work_cancel_delayable(&client->notify_work);
+#endif /* CONFIG_BT_HAS_PRESET_SUPPORT || CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+
+	bt_conn_get_info(client->conn, &info);
```
Should we add error check for `bt_conn_get_info`?

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -497,22 +755,25 @@ static int control_point_send_all(struct net_buf_simple *buf)
 
-		if (!client->conn) {
+		for (size_t j = 0U; j < ARRAY_SIZE(has_client_list); j++) {
+			if (has_client_list[i].context == context) {
+				client = &has_client_list[i];
```
```suggestion
			if (has_client_list[j].context == context) {
				client = &has_client_list[j];
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -601,77 +845,189 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
+		.properties = 0x00,
+		.name = " ",
+	};
+	int err;
 
-		preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
-			       preset_found, &next);
+	prev = preset_get_tail();
 
-		is_last = next == NULL;
+	err = bt_has_cp_generic_update(client, prev ? prev->index : 0x00, last.index,
```
`properties = 0x00` seems to have a special meaning. Do we have a define for the `0x00` property? Else suggest to add that.

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -601,77 +845,189 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
+	struct has_preset last = {
+		.index = client->context->last_preset_index_known,
+		.properties = 0x00,
+		.name = " ",
+	};
+	int err;
 
-		preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
-			       preset_found, &next);
+	prev = preset_get_tail();
 
-		is_last = next == NULL;
+	err = bt_has_cp_generic_update(client, prev ? prev->index : 0x00, last.index,
```
This could benefit from a comment stating why we are using `0x00` and `" "` 

Why are we using `" "` instead of `""`?

MariuszSkamra replied:
As per spec, the name has to be at least 1 byte long (1-40 bytes). I've changed it to "N/A" (not applicable/available)

Thalley replied:
> As per spec, the name has to be at least 1 byte long (1-40 bytes). I'll change it to "N/A" (not applicable/available)

Does that include the NULl terminator? Is the empty string (`" "`) allowed as a name?

MariuszSkamra replied:
> > As per spec, the name has to be at least 1 byte long (1-40 bytes). I'll change it to "N/A" (not applicable/available)
> 
> Does that include the NULl terminator? Is the empty string (`" "`) allowed as a name?

The NULL termiantor is not sent anyway. Regarding empty string, I'm not sure

Thalley replied:
Fair :)

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -601,77 +845,189 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
+
+	preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST, preset_found, &next);
+
+	is_last = next == NULL;
+
+	/* It is last Preset Changed notification if there are no presets left to notify and the
+	 * currently notified preset have the highest index known to the client.
+	 */
+	is_last = next == NULL && preset->index >= client->context->last_preset_index_known;
```
`is_last` assigned twice here

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -157,32 +156,35 @@ static struct bt_gatt_attr has_attrs[] = {
-#define ACTIVE_PRESET_INDEX_ATTR &has_attrs[7]
-#endif /* CONFIG_BT_HAS_PRESET_SUPPORT */
-#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+static struct bt_gatt_attr *hearing_aid_features_attr;
+static struct bt_gatt_attr *preset_control_point_attr;
+static struct bt_gatt_attr *active_preset_index_attr;
```
Should these be stored in `static struct bt_has has;`?

MariuszSkamra replied:
As `struct bt_has` is common for client and server I would not put it there as it would be unused by client.

MariuszSkamra commented on subsys/bluetooth/audio/has_internal.h:
```diff
@@ -46,6 +46,12 @@
+#define BT_HAS_ATT_NTF_HDR_LEN   3
+
+/* Minimum required ATT MTU size */
+#define BT_HAS_ATT_MTU_MIN      49
```
True

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -9,6 +9,7 @@ menuconfig BT_HAS
 	select EXPERIMENTAL
 	select UTF8
 	select BT_GATT_DYNAMIC_DB
+	select BT_GATT_AUTO_UPDATE_MTU
```
Can you elaborate the Android MTU topic? Was the MTU requested too small?

MariuszSkamra replied:
> This is an interesting addition.
> 
> We've seen issues with Android, where Android (as the central) did not properly update the MTU on connection, which mean that some notifications sent by us (as peripheral) could not be done, due to MTU not matching the requirements from e.g. BAP.

Can you elaborate? Was the MTU requested too small?

Thalley replied:
Yes, we've seen issues where we try to send notifications on reconnect of size e.g. 30 (assuming that the MTU is 64 as mandated by e.g. the BAP spec), because the phone did not request an update to the MTU before our first notification. I think it would make sense for all services in profiles that depend on BAP (CAP, TMAP, HAP, etc.) wait with sending notifications on reconnect until after the GATT MTU has been exchanged (at least the ones where it may be larger than the initial MTU of 23). 

I think that we should probably add `select BT_GATT_AUTO_UPDATE_MTU` to any profile enabled where we act as the GATT client (e.g. `BT_HAS_CLIENT`), but we should not enable it for profiles where we are the GATT server (e.g. `BT_HAS`), as we cannot be sure that `BT_GATT_CLIENT` is enabled. 

I am unsure what happens if a Kconfig `select`s another Kconfig without satisfying the dependencies.

MariuszSkamra replied:
> Yes, we've seen issues where we try to send notifications on reconnect of size e.g. 30 (assuming that the MTU is 64 as mandated by e.g. the BAP spec), because the phone did not request an update to the MTU before our first notification. I think it would make sense for all services in profiles that depend on BAP (CAP, TMAP, HAP, etc.) wait with sending notifications on reconnect until after the GATT MTU has been exchanged (at least the ones where it may be larger than the initial MTU of 23).

I would say it's not phone's fault that the server wants to send the notification that does not fit MTU.
On the other hand the server does not know when the client is ready to receive the notifications (e.g. the profiles or some application is ready).
As there is no dedicated L2CAP channel PSM connection, but ATT which is available right after connection this causes issues like that IMO.

> I think that we should probably add `select BT_GATT_AUTO_UPDATE_MTU` to any profile enabled where we act as the GATT client (e.g. `BT_HAS_CLIENT`), but we should not enable it for profiles where we are the GATT server (e.g. `BT_HAS`), as we cannot be sure that `BT_GATT_CLIENT` is enabled.
> 

`BT_HAS_CLIENT` selects `BT_GATT_AUTO_UPDATE_MTU` already. For profiles that are written on top of GATT Server I think that we should truncate the notifications, and apply some timeout waiting tor ATT exchange.

BTW, it would be helpful if we could have a function that will return true/false to know whether MTU was or was not already exchanged.

Thalley replied:
> BTW, it would be helpful if we could have a function that will return true/false to know whether MTU was or was not already exchanged.

We have `bt_gatt_get_mtu`, and I guess you could simply do `bool mtu_exchanged = bt_gatt_get_mtu() == BT_ATT_DEFAULT_LE_MTU`

Thalley replied:
Good point. 

IIRC the `bt_gatt_cb.att_mtu_updated` is even called on connection (`bt_att_connected` calls `att_chan_mtu_updated`), so we can't even use the callback. 

I'll raise a GH issue for this.

Thalley replied:
I've created https://github.com/zephyrproject-rtos/zephyr/issues/64175 and https://github.com/zephyrproject-rtos/zephyr/issues/64172

fredrikdanebjer commented on subsys/bluetooth/audio/has_internal.h:
```diff
@@ -46,6 +46,9 @@
 #define BT_HAS_IS_LAST 0x01
 
+/* ATT notification/indication header length (opcode + handle) */
+#define BT_HAS_ATT_NTF_HDR_LEN   3
```
I'm not sure what I think about this. You're basically defining the length of ATT_HANDLE_VALUE_NTF/ATT_HANDLE_VALUE_IND here, right?
I think that such a definition should reside in GATT probably. Since you are only using it once, consider doing it as has been done in other places, i.e. mark with a comment as such:
`/* PDU structure is [Opcode (1)] [Handle (2)] [...] */`

MariuszSkamra replied:
Yes, the structures are defined in `att_internal.h` (which I do not want to include).
I have a feeling that if I would did like you suggested then I would get a comment to not use magic numbers.
I'm OK with having a comment

MariuszSkamra replied:
done

Thalley replied:
This is a common issue for services: In order to properly truncate notifications, they need the notification header size. 

We should probably add that to `att.h`

fredrikdanebjer replied:
> Yes, the structures are defined in `att_internal.h` (which I do not want to include). I have a feeling that if I would did like you suggested then I would get a comment to not use magic numbers. I'm OK with having a comment

I'm also not a fan of magic numbers :D But seeing how widespread the use of them are in Zephyr, I think that the solution you did now is better regardless :)

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/61980#discussion_r1366690682

MariuszSkamra committed:
Bluetooth: audio: has: Truncate Control Point notifications to ATT MTU

This will truncate ATT notifications/indications if exceed ATT MTU size.
It is up to the client to exchange MTU.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Remove include of conn_internal header

This makes use of bt_conn_get_info function to access the conn address.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: has: Minor logging improvement

This adds more logs for debugging purposes.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
